controllers/argocd/notifications.go

```diff
@@ -551,3 +806,140 @@ func getNotificationsResources(cr *argoproj.ArgoCD) corev1.ResourceRequirements
+		}
+	}
+	return nil
+}
+
+// isNotificationsEnabled returns true if notifications are configured and enabled in the ArgoCD CR
+func isNotificationsEnabled(cr *argoproj.ArgoCD) bool {
+	return !reflect.DeepEqual(cr.Spec.Notifications, argoproj.ArgoCDNotifications{}) && cr.Spec.Notifications.Enabled
+}
+
+// cleanupUnmanagedNotificationsSourceNamespaceResources removes the notifications resources from target namespace
+func (r *ReconcileArgoCD) cleanupUnmanagedNotificationsSourceNamespaceResources(cr *argoproj.ArgoCD, ns string) error {
+	namespace := corev1.Namespace{}
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: ns}, &namespace); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	// Delete notifications role & rolebinding
+	existingRole := rbacv1.Role{}
+	roleName := getResourceNameForNotificationsSourceNamespaces(cr)
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: roleName, Namespace: namespace.Name}, &existingRole); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to fetch the role for the service account associated with %s : %s", common.ArgoCDNotificationsControllerComponent, err)
+		}
+	}
+	if existingRole.Name != "" {
+		argoutil.LogResourceDeletion(log, &existingRole, "cleaning up unmanaged notifications resources")
+		err := r.Delete(context.TODO(), &existingRole)
+		if err != nil {
+			return err
+		}
+	}
+
+	existingRoleBinding := &rbacv1.RoleBinding{}
+	roleBindingName := getResourceNameForNotificationsSourceNamespaces(cr)
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: roleBindingName, Namespace: namespace.Name}, existingRoleBinding); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get the rolebinding associated with %s : %s", common.ArgoCDNotificationsControllerComponent, err)
+		}
+	}
+	if existingRoleBinding.Name != "" {
+		argoutil.LogResourceDeletion(log, existingRoleBinding, "cleaning up unmanaged notifications resources")
+		if err := r.Delete(context.TODO(), existingRoleBinding); err != nil {
+			return err
+		}
+	}
+
+	// Delete NotificationsConfiguration CR in source namespace
+	notifCfg := &v1alpha1.NotificationsConfiguration{}
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: DefaultNotificationsConfigurationInstanceName, Namespace: namespace.Name}, notifCfg); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get the NotificationsConfiguration in namespace %s : %s", namespace.Name, err)
+		}
+	} else {
+		argoutil.LogResourceDeletion(log, notifCfg, "cleaning up unmanaged notifications resources")
+		if err := r.Delete(context.TODO(), notifCfg); err != nil {
+			return fmt.Errorf("failed to delete the NotificationsConfiguration in namespace %s : %s", namespace.Name, err)
+		}
+	}
+
+	// app-in-any-ns code will handle removal of notifications permissions for argocd-server in target namespace
```

Comment:
Do we add any permissions to argocd-server resources for notifications in any namespace?